kafka/signed_token_redeem_handler.go

```diff
@@ -151,25 +214,58 @@ func SignedTokenRedeemHandler(
 			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
 				Issuer_name:     "",
 				Issuer_cohort:   0,
-				Status:          redeemUnverified,
+				Status:          avroSchema.RedeemResultStatusUnverified,
 				Associated_data: request.Associated_data,
 			})
 			continue
 		} else {
-			logger.Trace().Msgf("request %s: validated", tokenRedeemRequestSet.Request_id)
+			logger.Info().Msg(fmt.Sprintf("request %s: validated", tokenRedeemRequestSet.Request_id))
 		}
-		if err := server.RedeemToken(verifiedIssuer, &tokenPreimage, request.Binding); err != nil {
-			logger.Error().Err(fmt.Errorf("request %s: token redemption failed: %w",
-				tokenRedeemRequestSet.Request_id, err)).
-				Msg("signed token redeem handler")
+		redemption, equivalence, err := server.CheckRedeemedTokenEquivalence(verifiedIssuer, &tokenPreimage, string(request.Binding), msg.Offset)
+		if err != nil {
+			message := fmt.Sprintf("request %s: failed to check redemption equivalence", tokenRedeemRequestSet.Request_id)
+			processingResult, errorResult := avroRedeemErrorResultFromError(
+				message,
+				msg,
+				producer,
+				tokenRedeemRequestSet.Request_id,
+				int32(avroSchema.RedeemResultStatusError),
+				log,
+			)
+			MayEmitIfPermanent(processingResult, errorResult, producer, log)
+			return NilIfPermanent(errorResult)
+		}
+
+		// Continue if there is a duplicate
+		switch equivalence {
+		case cbpServer.IDEquivalence:
+			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
+				Issuer_name:     "",
```

Comment:
should name and cohort be set? I know current behavior is to not but...

Reply:
Good idea.